grudge/trace_pair.py

```diff
@@ -0,0 +1,361 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+from arraycontext import (
+    ArrayContainer,
+    map_array_container,
+    with_container_arithmetic,
+    dataclass_array_container
+)
+
+from dataclasses import dataclass
+
+from numbers import Number
+
+from pytools import memoize_on_first_arg
+from pytools.obj_array import obj_array_vectorize, make_obj_array
+
+from meshmode.dof_array import flatten, unflatten
+from meshmode.mesh import BTAG_PARTITION
+
+import numpy as np
+import grudge.dof_desc as dof_desc
+
+
+__doc__ = """
+Trace Pairs
+^^^^^^^^^^^
+
+Container class
+---------------
+
+.. autoclass:: TracePair
+
+Boundary trace functions
+------------------------
+
+.. autofunction:: bdry_trace_pair
+.. autofunction:: bv_trace_pair
+
+Interior and cross-rank trace functions
+---------------------------------------
+
+.. autofunction:: interior_trace_pairs
+.. autofunction:: cross_rank_trace_pairs
+"""
+
+
+# {{{ Trace pair container class
+
+@with_container_arithmetic(
+    bcast_obj_array=False, eq_comparison=False, rel_comparison=False
+)
+@dataclass_array_container
+@dataclass(init=False, frozen=True)
+class TracePair:
+    """A container class for data (both interior and exterior restrictions)
+    on the boundaries of mesh elements.
+
+    .. attribute:: dd
+
+        an instance of :class:`grudge.dof_desc.DOFDesc` describing the
+        discretization on which :attr:`int` and :attr:`ext` live.
+
+    .. autoattribute:: int
+    .. autoattribute:: ext
+    .. autoattribute:: avg
+
+    .. note::
+
+        :class:`TracePair` is currently used both by the symbolic (deprecated)
+        and the current interfaces, with symbolic information or concrete data.
+    """
+
+    dd: dof_desc.DOFDesc
+    interior: ArrayContainer
+    exterior: ArrayContainer
+
+    def __init__(self, dd, *, interior, exterior):
+        object.__setattr__(self, "dd", dof_desc.as_dofdesc(dd))
+        object.__setattr__(self, "interior", interior)
+        object.__setattr__(self, "exterior", exterior)
```

Comment:
This might look crazy at first glance, but you can't do `self.dd = ...` (etc) for frozen `dataclasses`. I wanted to keep the constructor here (hence `init=False` in the `@dataclass` decorator). Ref: https://docs.python.org/3/library/dataclasses.html#frozen-instances